github/github.go

```diff
@@ -82,8 +87,11 @@ func (c *Config) NewClient() (*Client, error) {
 		transport = tr
 	}
 
-	transport = metrics.Transport{Transport: transport}
-	httpClient := &http.Client{Transport: transport}
+	cached := httpcache.NewTransport(httpcache.NewMemoryCache())
```

Comment:
@TingluoHuang Not exactly. But I'd say 90% of requests are cached :)

I've added a new log level `-3` so see if each request to GitHub API is cached or not. To enable the log level of `-3` you would write values.yaml like...

```
logLevel: "-3"
githubWebhookServer:
  logLevel: "-3"
```

Tail the logs from ARC controllers...

```
kubectl -n actions-runner-system logs -l app.kubernetes.io/name=actions-runner-controller -f | tee arc.log
```

And extract logged HTTP requests:

```
tail -f -n +1 arc.log | jq -rRc 'fromjson? | select(.from_cache != null) | "\(.ratelimit_remaining) \(.method) \(.url)"'
```

You'll see:

```
null GET https://api.github.com/orgs/$ORG/$REPO/actions/runners?per_page=100
null GET https://api.github.com/orgs/$ORG/$REPO/actions/runners?per_page=100
null GET https://api.github.com/orgs/$ORG/$REPO/actions/runners?per_page=100
3548 DELETE https://api.github.com/orgs/$ORG/$REPO/actions/runners/591
null GET https://api.github.com/orgs/$ORG/$REPO/actions/runners?per_page=100
null GET https://api.github.com/orgs/$ORG/$REPO/actions/runners?per_page=100
null GET https://api.github.com/orgs/$ORG/$REPO/actions/runners?per_page=100
null GET https://api.github.com/orgs/$ORG/$REPO/actions/runners?per_page=100
null GET https://api.github.com/orgs/$ORG/$REPO/actions/runners?per_page=100
3546 DELETE https://api.github.com/orgs/$ORG/$REPO/actions/runners/595
null GET https://api.github.com/orgs/$ORG/$REPO/actions/runners?per_page=100
null GET https://api.github.com/orgs/$ORG/$REPO/actions/runners?per_page=100
null GET https://api.github.com/orgs/$ORG/$REPO/actions/runners?per_page=100
3544 DELETE https://api.github.com/orgs/$ORG/$REPO/actions/runners/589
```

The first column is the `x-ratelimit-remaining` header value returned by GitHub API that is logged only when it was NOT served from cache.

You can see that the list repository runners API are always cached(the cache-controller header for that response says `max-age=60s` so httpcache caches it for 60 seconds so all the runners for a repository shares the same cache), and  it occasionally calls the Delete Runner API.

Reply:
🆒 
will 90% still hold when the large customer using ARC to scale hundreds of ephemeral runners since the response from the ListRunner endpoint should change everytime on a busy repo/org/enterprise?

Reply:
I guess it would scale well as this results in only one ListRunners call per 60s per RunnerDeployment.
So if a company has 1 enterprise runnerdeployment and 2 organizational runnerdeployments and 3 repository runnerdeployments, it might result in 6 calls per 60s.

Note that you only see a change in runner status (offline vs online, busy or not, runner existence, etc) only after a ListRunners API call after cache invalidation i.e. it will take up to 60s or so until you see changes in runner statuses.

Reply:
FWIW I think this is acceptable. The default cache control for github is 60s as you said, and it sounds acceptable to have a bit of leeway of 1min to detect an idle runner. The `duration` parameter protects the runners from being decommissioned too quickly assuming it's greater than the cache control so I can't see any downside in this approach afaict.

Reply:
> The duration parameter protects the runners from being decommissioned too quickly assuming it's greater than the cache control so I can't see any downside in this approach afaict.

Definitely! You should have `HRA.spec.scaleTrigggers[].duration` that is long enough to cover your longest-running workflow jobs anyway to let ARC not scale down the added runner too early in case of ARC failed to receive a webhook event of workflow_job completion so 60s of cache won't have real downside.

Reply:
@tumoa and I have been discussion about a potential enhancement on how scale trigger duration works at:

https://github.com/actions-runner-controller/actions-runner-controller/issues/911#issuecomment-1049715625